src/openlcb/IfTcpImpl.hxx

```diff
@@ -0,0 +1,553 @@
+    /// @return the OpenLCB priority in range 0..3 or uint_max if the message
+    /// is broken.
+    static unsigned guess_priority(const string &tcp_payload)
+    {
+        if (tcp_payload.size() < ABS_MTI_OFS + 2)
+        {
+            return UINT_MAX;
+        }
+        auto mti = (Defs::MTI)data_to_error(tcp_payload.data() + ABS_MTI_OFS);
+        return Defs::mti_priority(mti);
+    }
+
+    enum
+    {
+        FLAGS_OPENLCB_MSG = 0x8000,
+        FLAGS_CHAINING = 0x4000,
+        FLAGS_RESVD1_ZERO_CHECK = 0x3000,
+        FLAGS_FRAGMENT_NOT_FIRST = 0x0800,
+        FLAGS_FRAGMENT_NOT_LAST = 0x0400,
+        FLAGS_RESVD2_IGNORED = 0x03FF,
+
+        HDR_FLAG_OFS = 0,
+        HDR_SIZE_OFS = 2,
+        HDR_GATEWAY_OFS = 2 + 3,
+        HDR_SIZE_END = HDR_GATEWAY_OFS,
+        HDR_TIMESTAMP_OFS = 2 + 3 + 6,
+        HDR_LEN = 2 + 3 + 6 + 6,
+
+        MSG_MTI_OFS = 0,
+        MSG_SRC_OFS = 2,
+        MSG_DST_OFS = 2 + 6,
+        MSG_ADR_PAYLOAD_OFS = 2 + 6 + 6,
+
+        MSG_GLOBAL_PAYLOAD_OFS = MSG_DST_OFS,
+
+        MIN_MESSAGE_SIZE = MSG_GLOBAL_PAYLOAD_OFS + HDR_LEN,
+        MIN_ADR_MESSAGE_SIZE = MSG_ADR_PAYLOAD_OFS + HDR_LEN,
+
+        ABS_MTI_OFS = HDR_LEN + MSG_MTI_OFS,
+    };
+
+private:
+    /// No usable constructor; this is a static-only class.
+    TcpDefs();
+};
+
+/**
+   Virtual clock interface. Implementations are not required to be thread-safe.
+*/
+class SequenceNumberGenerator : public Destructable
+{
+public:
+    /** Returns the next strictly monotonic sequence number. */
+    virtual long long get_sequence_number() = 0;
+};
+
+/**
+   Implementation of sequence number generator that uses the real clock. Not
+   thread-safe.
+ */
+class ClockBaseSequenceNumberGenerator : public SequenceNumberGenerator
+{
+public:
+    long long get_sequence_number() override
+    {
+        long long current_time_ms = os_get_time_monotonic() / 1000000;
+        if (current_time_ms > sequence_)
+        {
+            sequence_ = current_time_ms;
+        }
+        else
+        {
+            ++sequence_;
+        }
+        return sequence_;
+    }
+
+private:
+    /// Sequence number of last sent message.
+    long long sequence_ = 0;
+};
+
+/**
+ * This flow renders outgoing OpenLCB messages to their TCP stream
+ * representation.
+ */
+class TcpSendFlow : public MessageStateFlowBase
+{
+public:
+    TcpSendFlow(If *service, NodeID gateway_node_id,
+        HubPortInterface *send_target, HubPortInterface *skip_member,
+        SequenceNumberGenerator *sequence)
+        : MessageStateFlowBase(service)
+        , sendTarget_(send_target)
+        , skipMember_(skip_member)
+        , gatewayId_(gateway_node_id)
+        , sequenceNumberGenerator_(sequence)
+    {
+    }
+
+private:
+    Action entry() override
+    {
+        auto id = nmsg()->dst.id;
+        if (id)
+        {
+            auto *n = iface()->lookup_local_node(id);
+            if (n)
+            {
+                // Addressed message loopback.
```

Comment:
This is interesting.  I would have expected loopback messages to happen at the general level and not the TCP specific level.  If that is the case, should this be an assert instead?

Reply:
We don't have a general level send flow at this point. The API of `If` gives you a send flow pointer, but that is a flow specific to the interface today, and there is no shared implementation (at this point) between different interfaces.
Maybe when we get to do the big refactoring where local interface and router/gateway are separated, then this functionality will go to the local interface instead.